src/misc/env.3
.\"(c) Copyright 1992 by Panagiotis Tsirigotis
.\"All rights reserved.  The file named COPYRIGHT specifies the terms
.\"and conditions for redistribution.
.\"
.\" $Id: env.3,v 1.2 2003/06/17 05:10:52 seth Exp $
.TH ENV 3L "20 October 1992"
.SH NAME
env_create, env_destroy, env_make, env_addvar, env_addstr, env_remvar, env_lookup, env_getvars -- environment manipulation functions
.SH SYNOPSIS
.LP
.nf
.ft B
#include "env.h"
.LP
.ft B
env_h env_create( env )
env_h env ;
.LP
.ft B
void env_destroy( env )
env_h env ;
.LP
.ft B
env_h env_make( env_strings )
char **env_strings ;
.LP
.ft B
int env_addvar( env, from_env, var )
env_h env ;
env_h from_env ;
char *var ;
.LP
.ft B
int env_addstr( env, str )
env_h env ;
char *str ;
.LP
.ft B
int env_remvar( env, var )
env_h env ;
char *var ;
.LP
.ft B
char **env_getvars( env )
env_h env ;
.SH DESCRIPTION
This library handles environments. An environment is a set of strings
of the form
.I "name=value".
In the following, we will use the term string as a synonym of
NUL-terminated array of
.I char.
.LP
.B env_create()
creates a new environment. The new environment will be empty unless
the argument
.I env
is not
.SB ENV_NULL.
In that case, the new environment will be a duplicate of
.I env
(i.e. they will contain the same strings).
.LP
.B env_destroy()
destroys the specified environment.
.LP
.B env_make()
creates a new environment which includes the
.I env_strings.
.I env_strings
should be a NULL-terminated array of strings.
.LP
.B env_addvar()
adds the specified variable
.I var
to
.I env.
The variable value is obtained from the environment
.I from_env.
If the variable exists already in
.I env
the old value is replaced with the new value.
.LP
.B env_addstr()
adds a string of the form
.I "name=value"
to
.I env.
.LP
.B env_remvar()
removes the specified variable
.I var
from the environment
.I env.
.LP
.B env_lookup()
searches
.I env
for variable
.I var.
It returns a string of the form
.I "name=value"
where
.I name
is the name of the variable
(i.e. it is equal to
.I var).
.LP
.B env_getvars
returns a NULL-terminated array of strings of the form
.I "name=value".
.SH "RETURN VALUES"
In case of error, all calls will place an error code in the global variable
.I env_errno.
Possible error codes:
.TP 15
.SB ENV_ENOMEM
out of memory
.TP
.SB ENV_EBADVAR
variable is not in environment
.TP
.SB ENV_EBADSTRING
string is not well-formed (i.e. is not of the form \fIname=value\fR).
.LP
.B env_create()
returns a handle or
.SM ENV_NULL
if it fails.
.LP
.B env_make()
returns a handle or
.SM ENV_NULL
if it fails.
.LP
.B env_addvar()
returns
.SM ENV_OK
on success or
.SM ENV_ERR
on failure.
.LP
.B env_addstr()
returns
.SM ENV_OK
on success or
.SM ENV_ERR
on failure.
.LP
.B env_remvar()
returns
.SM ENV_OK
on success or
.SM ENV_ERR
if the variable is not part of the environment.
.LP
.B env_loopkup()
returns a string on success or
.SM NULL
on failure.
.SH "SEE ALSO"
environ(5)


ChangeLog
2003-07-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (_fsm_free): Prevent infinite loops on fsm_free
	of corrupted memory

2003-07-09  James Tanis  <jtt@hawking.sysdetect.com>

	* src/dict/dll.c (dll_error_reason): Export pulibcally

2003-06-13  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h, include/ht.h, include/pq.h: avoid stupid void * #1311

2003-06-07  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_create): argsp may change size between
	external and internal users due to HASH_STATS define.

2003-05-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_destroy): Allow clean up of all memory when
	insure is in use and fsma is replaced by malloc

2003-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_STATS): keep performance stats for hash tables

	* src/dict/htimpl.h (struct ht_stats): hash table performance stats
	(ENTRY_IS_FULL): kludge to detect bucket collisions

	* src/dict/ht.c: instrumented to keep performance stats for hash tables
	(print_hashstats): print performance stats

2003-04-28  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/ht.c (ht_delete): microscopic performance tweak
	[Seth got the big win already]

2003-04-16  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Only allow users to
	actually alter the value of thread_overried if we are compiled
	with threads.

2003-04-08  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Provide interface for
	people to specify this will be a threaded program and to enforce
	thread safe memory protection.

2003-04-03  Seth Robertson  <seth@startide.sysdetect.com>

	* src/pq/hpq.c: Make thread-safe, through PQ_THREADED_SAFE

2003-03-31  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/*.c src/fsma/*.c: Make thread-safe, through
	DICT_THREADED_SAFE and DICT_THREADED_MEMORY

2002-12-10  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/pq/Makefile: move -lpq to -lpqueue so as not to conflict
	with postgres

2002-11-11  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/fsma/fsma.c, include/dict.h: opx_rollup_3 merge

2002-11-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* include/dict.h, src/fsma/fsma.c: trheads merge

2002-11-06  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/fsma/fsma.c: Compile problems when using dmalloc

2002-11-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* include/dict.h: Define symbol for NOCOALESCE which makes it
	clear the only reason we are using it is because we are too lazy
	to make CLC thread-safe

2002-09-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/dict/ht.c: insight nits

2002-09-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/dict/dll.c, ChangeLog: Do not create dlls w/o UNORDERED and
	w/o OOCMP function--coredump otherwise sometimes

2002-09-05  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/dll.c (dll_create): Default order is ordered.

	* src/dict/dll.c (dll_create): Use bitwise and for bits, instead
	of addition.  Prevented an argument sanity check from occurring.

2002-08-29  Seth Robertson  <seth@startide.sysdetect.com>

	* include/fsma.h (struct __fsma_header): AHHHHH-HHHHAAAAAAAA!
	The reference counter was previously a short.  Worker, for reasons
	too complicated to get into, can have more than 2^16th lists.
	If we happen (not unlikely) to destroy down to 2^16 (e.g. zero)
	fsma thinks all allocators are free and promptly proceeds to return
	the memory to the system.  Memory was free'd but still in use.
	BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA.

	* src/fsma/fsma.c (fsm_destroy): Perform assertion to verify we
	actually found the fsma allocator in the coalesce list.  Don't
	bother to realloc the coalesce list size down--it won't do no good
	nohow.

2002-08-27  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Serious and quite possibly
	bug-injecting misuse of a hint (cur_min).  The insert and delete
	routines were setting this to a bucket chain entry, not a table
	entry bucket (except when the insert routine had to perform a
	bucket chain expansion--when cur_min might be set to some other
	random value).  This usually was OK since bucket chain numbers are
	very low (0-3 range), but if a valid entry was in the table at
	slot zero or one, then ht_minimum might never see that item.

2002-08-14  James Tanis  <jtt@saint.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Use cached minimum index as search start.
	(ht_do_insert): Update minimum index if less the current minimum.
	(ht_delete): Search for new minimum index if the index of the
	deleted node equals the current minimum.
	In addition add and obj_cnt so we don't try to perform expensive
	operations on empty hashes.

2002-05-13  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): Shared libfsma no longer
	breaks with libtool (on both linux and bsd).

2002-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): back out shared libfsma;
	this breaks libtool in a strange way that requires investigation

2002-05-10  James Tanis  <jtt@winwood.sysdetect.com>

	* src/fsma/fsma.c: Formatting and simplify coalesce code.

	* src/fsma/Makefile (BK_LARGE_LIB): Make libfsma shared.

	* src/dict/bst.c (bst_redelete): Improve efficiency by avoiding
	unnecessary checks and recusive calls.

2002-02-22  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_NUKE): delete contents and dict itself
	(_DICT_NUKE_WORK_LOOP): common code for DICT_NUKE, DICT_NUKE_CONTENTS

	* src/dict/dictimpl.h (HANDLE_ERROR): Eliminate unused id argument
	[use BK_FUNCNAME  here instead if you need it for debugging]

2002-01-31  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/Makefile (BK_LARGE_LIB): use ${LIBEXT} for libtool build

2002-01-09  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/fsma.c: INSIGHT->INSURE


src/misc/env.c
/*
 * (c) Copyright 1992 by Panagiotis Tsirigotis
 * All rights reserved.  The file named COPYRIGHT specifies the terms 
 * and conditions for redistribution.
 */

static char RCSid[] = "$Id: env.c,v 1.1 2001/05/26 22:04:49 seth Exp $" ;

#include <stdlib.h>
#include <memory.h>
#include <string.h>

#include "misc.h"
#include "env.h"
#include "clchack.h"

typedef struct __env env_s ;

#define PRIVATE					static
#define INITIAL_VARS				20
#define INCREASE					10

#ifndef NULL
#define NULL						0
#endif

#if 0
#if !defined(__linux__) && !defined(__bsdi__)	/* do we really need this?... */
/* NO! It conflicts w/ std headers. -Erez, 7/27/00 */
char *malloc(size_t) ;
char *realloc(void *, size_t) ;
#endif
#endif

int env_errno ;

PRIVATE char **lookup(env_h env, char *var, register int len);


PRIVATE env_s *alloc_env(unsigned int max_vars)
{
	env_s *ep ;
	unsigned size ;
	char **pointers ;

	ep = (env_s *) malloc( sizeof( env_s ) ) ;
	if ( ep == ENV_NULL )
	{
		env_errno = ENV_ENOMEM ;
		return( ENV_NULL ) ;
	}

	size = ( max_vars + 1 ) * sizeof( char * ) ;
	pointers = (char **) malloc( size ) ;
	if ( pointers == NULL )
	{
		free( (char *)ep ) ;
		env_errno = ENV_ENOMEM ;
		return( ENV_NULL ) ;
	}
	(void) memset( (char *)pointers, 0, (int) size ) ;

	ep->vars = pointers ;
	ep->max_vars = max_vars ;
	ep->n_vars = 0 ;
	return( ep ) ;
}


env_h env_create(env_h init_env)
{
	unsigned u ;
	env_s *ep ;
	unsigned max_vars ;

	if ( init_env == ENV_NULL )
		max_vars = INITIAL_VARS ;
	else
		max_vars = init_env->n_vars + 5 ;
	
	ep = alloc_env( max_vars ) ;
	if ( ep == NULL )
	{
		env_errno = ENV_ENOMEM ;
		return( ENV_NULL ) ;
	}

	if ( init_env == ENV_NULL )
		return( ep ) ;

	for ( u = 0, ep->n_vars = 0 ; u < init_env->n_vars ; u++, ep->n_vars++ )
	{
		ep->vars[ ep->n_vars ] = make_string( 1, init_env->vars[ u ] ) ;
		if ( ep->vars[ ep->n_vars ] == NULL )
		{
			env_destroy( ep ) ;
			env_errno = ENV_ENOMEM ;
			return( ENV_NULL ) ;
		}
	}
	return( ep ) ;
}


void env_destroy(env_h env)
{
	unsigned u ;

	for ( u = 0 ; u < env->n_vars ; u++ )
		free( env->vars[ u ] ) ;
	free( (char *)env->vars ) ;
	free( (char *)env ) ;
}


env_h env_make(char **env_strings)
{
	env_s *ep ;
	char **pp ;

	for ( pp = env_strings ; *pp ; pp++ ) ;

	ep = alloc_env( (unsigned) (pp-env_strings) ) ;
	if ( ep == NULL )
	{
		env_errno = ENV_ENOMEM ;
		return( ENV_NULL ) ;
	}

	for ( pp = env_strings ; *pp ; pp++ )
	{
		char *p = make_string( 1, *pp ) ;

		if ( p == NULL )
		{
			env_destroy( ep ) ;
			env_errno = ENV_ENOMEM ;
			return( ENV_NULL ) ;
		}
		ep->vars[ ep->n_vars++ ] = p ;
	}
	return( ep ) ;
}


char *env_lookup(env_h env, char *var)
{
	char **lookup() ;
	char **pp = lookup( env, var, strlen( var ) ) ;

	return( ( pp == NULL ) ? NULL : *pp ) ;
}


PRIVATE char **lookup(env_h env, char *var, register int len)
{
	register char **pp ;

	for ( pp = env->vars ; *pp ; pp++ )
		if ( strncmp( *pp, var, len ) == 0 && (*pp)[ len ] == '=' )
			return( pp ) ;
	return( NULL ) ;
}


PRIVATE int grow(env_s *ep)
{
	char **new_vars ;
	unsigned new_max_vars ;
	unsigned new_size ;

	new_max_vars = ep->max_vars + INCREASE ;
	new_size = ( new_max_vars+1 ) * sizeof( char * ) ;
	new_vars = (char **) realloc( (char *)ep->vars, new_size ) ;
	if ( new_vars == NULL )
		return( ENV_ERR ) ;
	
	ep->vars = new_vars ;
	ep->max_vars = new_max_vars ;
	return( ENV_OK ) ;
}


/*
 * Add the variable string to the given environment.
 */
PRIVATE int addstring(env_s *ep, char *var_string, int len)
{
	char **pp ;
	char *p ;

	p = make_string( 1, var_string ) ;
	if ( p == NULL )
		return( ENV_ERR ) ;
		
	pp = lookup( ep, var_string, len ) ;
	if ( pp == NULL )
	{
		if ( ep->n_vars >= ep->max_vars && grow( ep ) == ENV_ERR )
		{
			free( p ) ;
			env_errno = ENV_ENOMEM ;
			return( ENV_ERR ) ;
		}
		ep->vars[ ep->n_vars++ ] = p ;
	}
	else
	{
		free( *pp ) ;
		*pp = p ;
	}
	return( ENV_OK ) ;
}


int env_addvar(env_h env, env_h from_env, char *var_name)
{
	char *var_string = env_lookup( from_env, var_name ) ;

	if ( var_string == NULL )
	{
		env_errno = ENV_EBADVAR ;
		return( ENV_ERR ) ;
	}

	return( addstring( env, var_string, strlen( var_name ) ) ) ;
}


int env_addstr(env_h env, char *var_string)
{
	char *p = strchr( var_string, '=' ) ;

	if ( p == NULL )
	{
		env_errno = ENV_EBADSTRING ;
		return( ENV_ERR ) ;
	}
	
	return( addstring( env, var_string, p-var_string ) ) ;
}


int env_remvar(env_h env, char *var)
{
	char **pp = lookup( env, var, strlen( var ) ) ;

	if ( pp == NULL )
	{
		env_errno = ENV_EBADVAR ;
		return( ENV_ERR ) ;
	}
	
	free( *pp ) ;
	*pp = env->vars[ --env->n_vars ] ;
	return( ENV_OK ) ;
}


#ifdef notdef
PRIVATE int comparator( p1, p2 )
	char **p1, **p2 ;
{
	return( strcmp( *p1, *p2 ) ) ;
}


void env_sort( env )
	env_h env ;
{
	qsort( (char *)env->vars, env->n_vars, sizeof( char * ), comparator ) ;
}
#endif
